### PR TITLE
[Feature] Support clickhouse jdbc datasource(#40894)

### PR DESCRIPTION
**Pull request by DataScientistSamChan**
## Why I'm doing:
Enhance current clickhouse JDBC datasource support.

## What I'm doing:
Support more fine-grained clickhouse JDBC datasource with partition information detection and more fine grained data types.

Fixes #issue
#40894

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exec/jdbc_scanner.cpp [hunk 1]
```diff
@@ -261,13 +300,13 @@ StatusOr<LogicalType> JDBCScanner::_precheck_data_type(const std::string& java_c
...
         return TYPE_VARCHAR;
-    } else if (java_class == "java.sql.Date") {
+    } else if (java_class == "java.sql.Date" ) {
         if (type != TYPE_DATE) {
             return Status::NotSupported(
                     fmt::format("Type mismatches on column[{}], JDBC result type is Date, please set the type to date",
                                 slot_desc->col_name()));
         }
-        return TYPE_VARCHAR;
+        return TYPE_DATE;
```
Comment (zombee0) [inline be/src/exec/jdbc_scanner.cpp:309]: is this ok?

**Review by zombee0 — COMMENTED** [review]
clang-format

File: be/src/exec/jdbc_scanner.cpp (hunk 1 above)
Comment (DataScientistSamChan) [reply be/src/exec/jdbc_scanner.cpp]: Should be `TYPE_VARCHAR`, fixed.

File: fe/fe-core/src/main/java/com/starrocks/connector/jdbc/ClickhouseSchemaResolver.java [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+                        " WHERE database = ? " + "AND table = ? AND name IS NOT NULL " +
+                        "GROUP BY partition ORDER BY partition";
+        if (table.isUnPartitioned()) {
+            throw new StarRocksConnectorException("GetPartitionQuery for unpartitioned table is not implemented");
+        } else {
+            return partitionQuery;
+        }
+    }
+
+    public static void main(String[] args) {
```
Comment (zombee0) [inline fe/fe-core/src/main/java/com/starrocks/connector/jdbc/ClickhouseSchemaResolver.java:358]: testing code?

Comment (DataScientistSamChan) [reply fe/fe-core/src/main/java/com/starrocks/connector/jdbc/ClickhouseSchemaResolver.java:382]: Sorry about that, I am kind of using the git for synchronizing the code between local mac and my dev server, hence, the main for test. It's all cleaned up in the latest commit. With unittest added, and function working just fine in my environment.

**Review by wangsimo0 — DISMISSED** [review]

**Review by zombee0 — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/jdbc/ClickhouseSchemaResolver.java [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+                primitiveType = PrimitiveType.DATE;
+                break;
+            case "DateTime":
+                primitiveType = PrimitiveType.DATETIME;
+                break;
+            default:
+                // Decimal(9,9), first 9 is precision, second 9 is scale
+                if (typeName.startsWith("Decimal")) {
+                    String[] precisionAndScale =
+                            typeName.replace("Decimal", "").replace("(", "")
```
Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/connector/jdbc/ClickhouseSchemaResolver.java:198]: do you have to call `replaceAll`?

**Review by dirtysalt — DISMISSED** [review]

**Review by dirtysalt — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/jdbc/ClickhouseSchemaResolver.java (hunk 3 above)
Comment (DataScientistSamChan) [reply fe/fe-core/src/main/java/com/starrocks/connector/jdbc/ClickhouseSchemaResolver.java:143]: I am trying to extract the precision and scale here, and using `replace` is just a convenient way to get them. Another option would be using regular expression. What's the concern here?

**Review by Seaven — DISMISSED** [review]

**Review by Youngwb — DISMISSED** [review]

**Review by dirtysalt — DISMISSED** [review]

**Review by wangsimo0 — DISMISSED** [review]

**Review by wangsimo0 — APPROVED** [review]

**Comment by zombee0** [comment]
https://github.com/Mergifyio backport branch-3.3

**Pull request merged**